### ISSUE TITLE
Example queries - mcd

### ISSUE THREAD

**Issue by desaperados** [makerdao/vulcan.spec#10]
It came up on the weekly call today that we should explore some example queries with the frontend teams who will be using Vulcanize in order to ensure that we have a good shared understanding of the requirements. 

Here's a few of the queries from the schema which are relevant to retrieving Cdps for the upcoming milestones. Let's use these as a starting point for some concrete examples of queries which reflect frontend requirements:

```graphql
# Urns collection
allUrns(filters: {}): [Urn]

# Retrieve a single Urn
getUrn(
  ilkId: String!
  urnId: String!
  blockNumber: Int 
): Urn

# Aggregate Urns over a given interval
allUrnStates(
  filters: {}
  interval: String # optionally specify a postgres aggregation interval e.g '1 day'
): [Urn]

# Frob events collection
allFrobs(filters: {}): [FrobEvent]
```

@levity @rmulhol @elizabethengelman @chrisbradbury88 @wkampmann @m0ar

**Comment by m0ar** [comment]
@desaperados Checking my assumptions:
- `allUrns` returns a set of *all* urns at their current state
- `allUrnStates` returns a set of historical states for a *single* `urn`


Also, what's the intended behaviour of `updated` in historical states? The last poke _before_ the current change? :thinking:

**Comment by m0ar** [comment]
@desaperados @levity I see some time values in [mcd.graphql](https://github.com/makerdao/vulcan.spec/blob/master/mcd.graphql) are listed as `Datetime`, and some in the `Era` type. What's the reason some queries also need the epoch, but some exclusively human-readable format?

**Comment by desaperados** [comment]
@m0ar `Era` was intended to wrap block timestamps on transaction objects, providing both the seconds since the epoch as well as the convenience of a human readable datetime format. There's no reason why it couldn't also be applied to non-transaction objects for consistency if that was desired.

**Comment by levity** [comment]
For price history, we'll need a query that can fetch PokeEvents within a given time range.

**Comment by m0ar** [comment]
@rmulhol Do we have what @levity is asking for in the pipeline atm?

**Comment by rmulhol** [comment]
@levity @m0ar right now we're tracking the [`LogMedianPrice` event](https://github.com/makerdao/median/blob/master/src/median.sol#L101) that's emitted at the end of a successful `poke` invocation. If that works, we can create a story to write the query that would return a collection of those within a given time range - shouldn't be too difficult.

However, if we want to track all of the times that `poke` was called (including transactions that were reverted due to failed assertions, e.g. [this](https://github.com/makerdao/median/blob/master/src/median.sol#L69)), then I think we'd need the contracts team to emit an event earlier in the function's execution (perhaps a `note` event?).

Let me know which approach sounds better, or if you had something different in mind.

**Comment by levity** [comment]
> @levity @m0ar right now we're tracking the [`LogMedianPrice` event](https://github.com/makerdao/median/blob/master/src/median.sol#L101) that's emitted at the end of a successful `poke` invocation. If that works, we can create a story to write the query that would return a collection of those within a given time range - shouldn't be too difficult.

That sounds fine to me. I don't think we need to cover failed transactions.

**Comment by rmulhol** [comment]
One small request here - would it be possible to avoid using `Id` in field names for our queries/structs? It gets a bit confusing when we're working with internal database IDs to also have that name referring to addresses, strings, etc.

Some proposed alternatives:
- for ilk, we refer to a string like `ETH` as `ilkName` - maybe that could be applied here?
- urn is a bit trickier since the type might(?) still be up in the air, but we've sometimes used `urnGuy`

**Comment by levity** [comment]
Would it make sense to define a [custom scalar type](https://graphql.org/learn/schema/#scalar-types) "Address" to reduce ambiguity in these cases? I imagine it would just be the same as String in practice, but at least it would be a signal to users/developers. (Though maybe down the road you could add address validation.) Not sure if that's complicated to handle.

**Comment by m0ar** [comment]
We still need to name that field though, but I'm all for types :grin:

**Comment by desaperados** [comment]
Yeah agree that the issue of how to treat `Ilk` and `Urn` identifiers is a bit problematic. Would be nice to be able to just use `ilk::Address` and `urn::String` but that then raises the question of how to refer to the actual object in relational queries.

One approach would be to use `ilk` and `urn` to refer to identifiers and then use something like `getIlk` and `getUrn` to refer to actual objects:

```graphql
type Urn {
  urn: Address
  ilk: String
  getIlk: Ilk
}
```

The other approach is what we're doing now - I went with `Id` as the descriptor for object identifiers as it seemed least controversial but we could use another handle if `Id` is problematic. `Index` or `key` could be options. Not sure about `name` as that doesn't really convey the same meaning.

**Comment by m0ar** [comment]
@desaperados Wholeheartedly agree that naming and terminology is a bit of a mess atm. I think the gist of the issue is: what do we mean with the `ilk` string in that type? 

The frontend/js teams have requested the possibility to query by, and get decoded "ilk names", like `ETH` instead of ye olde `0x4554480000000000000000000000000000000000000000000000000000000000`. We've added the data architecture to support that, but whether or not `name` is a good `name` for it, I'm not sure :man_shrugging: 

Also, `urn` in the `Urn` is a bit confusing. What about 
```graphql
type Urn {
    guy: Address
    ilk: String (still an issue though)
    getIlk: Ilk
}
```
But on the other hand, all of these nested types is going to make the queries messier isn't it

**Comment by elizabethengelman** [comment]
Instead of using `id`, because it seems to imply a database generated integer id (for me at least 😄), I wonder if we could use the term `identifier` instead - a small change, but it may help with understanding? So we can keep the current approach and just sub `id` for `identifier`.

```graphql
type Urn {
  ilkIdentifier: String
  urnIdentifer: Address
  ilk: Ilk
}
```

**Comment by desaperados** [comment]
@elizabethengelman I quite like `identifier` 👍 

@m0ar I'd assumed that vdb would be exposing friendly human readable types rather than hexes for numbers, strings etc.

```graphql
getIlk(identifier: "ETH-A") 
```

However, this raises the question of how to handle bytecode representations when/if required.
